lib/inMemFileDB.js

```diff
@@ -0,0 +1,346 @@
+                    }
+                } else {
+                    this.dataset = {};
+                }
+            }
+        } else if (fs.existsSync(this.datasetName + '.bak')) {
+            try {
+                this.dataset = JSON.parse(fs.readFileSync(this.datasetName + '.bak').toString());
+            } catch (e) {
+                this.log.error(this.namespace + ' Cannot parse ' + this.datasetName + '.bak: ' + e);
+                this.dataset = {};
+            }
+        } else {
+            this.dataset = {};
+        }
+
+        // Check if directory exists
+        this.datasetName = this.datasetName.replace(/\\/g, '/');
+        /** @type {string|string[]} */
+        let parts = this.datasetName.split('/');
+        parts.pop();
+        parts = parts.join('/');
+        if (!fs.existsSync(parts)) fs.mkdirSync(parts);
+    }
+
+    handleSubscribe(client, type, pattern, options, cb) {
+        if (typeof options === 'function') {
+            cb = options;
+            options = undefined;
+        }
+        client._subscribe = client._subscribe || {};
+        if (!client._subscribe[type]) client._subscribe[type] = [];
+        const s = client._subscribe[type];
+
+        if (pattern instanceof Array) {
+            pattern.forEach(pattern => {
+                for (let i = 0; i < s.length; i++) {
+                    if (s[i].pattern === pattern) return;
+                }
+
+                s.push({pattern: pattern, regex: new RegExp(tools.pattern2RegEx(pattern)), options: options});
+            });
+        } else {
+            for (let i = 0; i < s.length; i++) {
+                if (s[i].pattern === pattern) {
+                    typeof cb === 'function' && cb();
+                    return;
+                }
+            }
+
+            s.push({pattern: pattern, regex: new RegExp(tools.pattern2RegEx(pattern)), options: options});
+        }
+        typeof cb === 'function' && cb();
+    }
+
+    handleUnsubscribe(client, type, pattern, cb) {
+        if (!client._subscribe || !client._subscribe[type]) {
+            if (typeof cb === 'function') cb();
+            return;
+        }
+        const s = client._subscribe[type];
+        if (pattern instanceof Array) {
+            pattern.forEach(pattern => {
+                for (let i = 0; i < s.length; i++) {
+                    if (s[i].pattern === pattern) {
+                        s.splice(i, 1);
+                        return;
+                    }
+                }
+            });
+        } else {
+            for (let i = 0; i < s.length; i++) {
+                if (s[i].pattern === pattern) {
+                    s.splice(i, 1);
+                    if (typeof cb === 'function') cb();
+                    return;
+                }
+            }
+        }
+        if (typeof cb === 'function') cb();
+
+    }
+
+    publishToClients(client, type, id, obj) {
+        throw new Error('no communication handling implemented');
+    }
```

Comment:
Does it make sense to define an "abstract" method here if it can only be used after being overwritten?

Reply:
In fact "throw new error" is best practice at least what I found

Reply:
I'm not sure if this even belongs into the DB class when it has no implementation and is only needed in the Server classes.

Reply:
Needs to because I also put the whole subscription (and so also publishing) logic in the main class because it is used for all. But how to publish to connected clients is specific to the communication implementation later

Reply:
Ok, makes sense! :+1: